pinc/upload_file.inc

```diff
@@ -279,20 +288,40 @@ function show_upload_message($verbose, $message)
 // which should be updated if the below logic changes.
 function is_valid_filename($filename, $restrict_extension=False)
 {
+    // Filenames can't be longer than 200 characters total
+    if(strlen($filename) > 200)
+    {
+        return false;
+    }
+
+    // If we want to restrict the filename extension, check that
+    if($restrict_extension !== FALSE && !endswith($filename, $restrict_extension))
+    {
+        return false;
+    }
+
     // Base the filename restrictions on removing anything that could
     // conceivably be a shell escape, control character etc.
     // See http://www.owasp.org/index.php/Unrestricted_File_Upload
-    if ($restrict_extension === False) {
-        // Ordinarly we allow filenames to start with an alphanumeric, followed
-        // by 0 or more alphanumerics, hypens, dashes or periods.
-        $regexp = '/^\w[\w\.-]{0,199}$/';
-    } else {
-        // If we want to restrict filename extensions, '.'s aren't allowed in the
-        // body of the filename, and the filename must end with '.$restrict_extension'
-        $limit = 199 - (strlen($restrict_extension) + 1);
-        $regexp = "/^\w+[\w-]{0,$limit}\." . $restrict_extension . '$/i';
+
+    // We allow filenames to start with an alphanumeric, followed by 0 more
+    // alphanumerics, dashes and ending with a period and an extension.
+    // The filename is valid if the regexp matches exactly once.
+    return preg_match('/^\w[\w-]*\.\w+$/', $filename) == 1;
+}
+
+// This function is used in remote_file_manager to enforce valid
+// directory names. It pairs with is_valid_filename() above.
+function is_valid_dirname($dirname)
+{
+    // Directory names can't be longer than 200 characters total
+    if(strlen($dirname) > 200)
+    {
+        return false;
     }
 
+    // We allow filenames to start with an alphanumeric, followed by 0 more
+    // alphanumerics, dashes, and periods.
```

Comment:
This is only a suggestion -- when I first saw the regex, I had to go look up `\w` to discover that the underscore is part of that character class. I don't think of the term "alphanumerics" as including the underscore, though. The current comment also says "periods", plural, which is incorrect.

```
// ... followed by 0 or more
// alphanumerics, underscores, hyphens and one dot separating
// the filename from the extension"
```

Reply:
I'll update to include underscores. For directories it is one or more periods, however.